qcore/geo.py

```diff
@@ -827,3 +830,300 @@ def compute_intermediate_latitudes(lon_lat1, lon_lat2, lon_in):
+    q = homogenise_point(q)
+    return projective_span(pi, p, q)
+
+
+def closest_points_between_planes(
+    p1_corners: np.ndarray, p2_corners: np.ndarray
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Compute the closest points between two finite planes.
+
+    This function solves the closest point problem by phrasing it as a
+    constrained quadratic optimisation problem, and then solving using scipy's
+    optimisation package.
+
+    Parameters
+    ----------
+    p1_corners : np.ndarray
+        The corners of the first plane.
+        p1_corners[i] and p1_corners[( i + 1 ) % 4] must be adjacent corners.
+    p2_corners : np.ndarray
+        The corners of the second plane.
+        p1_corners[i] and p1_corners[( i + 1 ) % 4] must be adjacent corners.
+
+    Returns
+    -------
+    (np.ndarray, np.ndarray)
+        The point pair (p, q) with p in p1 and q in p2 such that ||p - q|| is minimised.
+    """
+    # These dual coordinates are used to enforce the condition that the found
+    # points lie within p1 and p2 respectively.
+    p1_dual_coordinates = plane_from_three_points(*p1_corners[:3])
+    p2_dual_coordinates = plane_from_three_points(*p2_corners[:3])
+
+    p1_in_plane_constraint = {
+        "type": "eq",
+        "fun": lambda x: p1_dual_coordinates.dot(homogenise_point(x[:3])),
+    }
+
+    p2_in_plane_constraint = {
+        "type": "eq",
+        "fun": lambda x: p2_dual_coordinates.dot(homogenise_point(x[3:])),
+    }
+
+    p1_centroid = np.average(p1_corners, axis=0)
+    # For each side of the plane p1, we construct a plane orthogonal to p1
+    # passing through the side.
+    p1_ortho_planes = [
+        orthogonal_plane(
+            p1_dual_coordinates, p1_corners[i], p1_corners[(i + 1) % len(p1_corners)]
+        )
+        for i in range(len(p1_corners))
+    ]
+
+    # The dual coordinates of a plane are defined up to scalar multiples. Here we
+    # scale the coordinates to ensure that the normal vectors point towards the
+    # centre.
+    #
+    # The picture should look like
+    #
+    #      x
+    #     / \
+    # p1'/  |
+    #   /    x--------x
+    #  X  ---> norm  /
+    #   \  /   .    / p1
+    #   | /   c    /
+    #    x--------x
+    #
+    # Where p1 is the plane, p1' the plane orthogonal to p1, and norm the normal vector pointing towards c, the centroid.
+    # This ensures that the inequality norm * x >= 0 describes all points on the same side of p1' as c.
+    # Adding a condition like this for each edge of p1 bounds our search to just points on p1.
+    for i, norm in enumerate(p1_ortho_planes):
+        if norm.dot(homogenise_point(p1_centroid)) < 0:
+            p1_ortho_planes[i] *= -1
+
+    # We now repeat this process for p2.
+    p2_centroid = np.average(p2_corners, axis=0)
+    p2_ortho_planes = [
+        orthogonal_plane(
+            p2_dual_coordinates, p2_corners[i], p2_corners[(i + 1) % len(p2_corners)]
+        )
+        for i in range(len(p2_corners))
+    ]
```

Comment:
Since this is a repeated process could this and the explanation be it's own function and just done for both p1 and p2